tests/pre-exit-success.bats

```diff
@@ -57,6 +57,26 @@ COMMON_CURL_OPTIONS='--form \* --form \* --form \* --form \* --form \* --form \*
   assert_output --partial "curl success 3"
 }
 
+@test "Uploads multiple files concurrently" {
+  export BUILDKITE_PLUGIN_TEST_COLLECTOR_FILES='**/*/junit-*.xml'
+  export BUILDKITE_PLUGIN_TEST_COLLECTOR_UPLOAD_CONCURRENCY='3'
+
+  stub curl \
+    "-X POST --silent --show-error --max-time 30 --form format=junit ${COMMON_CURL_OPTIONS} \* -H \* : echo 'curl success 1'"
+
+  run "$PWD/hooks/pre-exit"
+
+  unstub curl
+
+  assert_success
+  assert_output --partial "Uploading './tests/fixtures/junit-1.xml'..."
+  assert_output --partial "Uploading './tests/fixtures/junit-2.xml'..."
+  assert_output --partial "Uploading './tests/fixtures/junit-3.xml'..."
+  assert_output --partial "curl success 1"
+  assert_output --partial "curl success 1"
+  assert_output --partial "curl success 1"
```

Comment:
Another alternative would be to actually count the line amounts:

```
  assert_equal "$(echo "$output" | grep -c "curl success")" "3"
```

Reply:
I implemented your alternative suggestion, I think that sufficiently covers the intent of the test.

Reply:
To fix the cross-platform compatibility issues I've replaced `ps -p` with `kill -0` it was the only test I could find that had consistent behaviour across platforms to test a pid exists.

Reply:
I've been attempting to add the missing tests, but I just keep seem to running into the error:

```
2023-06-19 06:37:02 UTC | ✗ Concurrency waits when the queue is full
-- | --
  | 2023-06-19 06:37:04 UTC | (from function `unstub' in file /usr/lib/bats/bats-mock/stub.bash, line 70,
  | 2023-06-19 06:37:04 UTC | in test file tests/pre-exit-success.bats, line 89)
  | 2023-06-19 06:37:04 UTC | `unstub curl' failed
```

I'm not very familiar with bats, it seems as though I'm likely configuring it wrong but bats-mock is hiding the underlying error?

Reply:
at the top of each test file there is a line showing [the variable you can set to get stub debugging information](https://github.com/buildkite-plugins/test-collector-buildkite-plugin/blob/main/tests/pre-exit-success.bats#L4). You can uncomment the line or set it when running the tester locally:

```
docker run --rm -ti -e CURL_STUB_DEBUG=/dev/tty -v $PWD:/plugin buildkite/plugin-tester:v4.1.0 bats tests/pre-exit-success.bats -f 'string on the test name to just run that one'
```

That said, my guess is that it is what I originally expected to happen with the tests: curl is being run more than once, so it needs to be stubbed that amount of times. This was previously hidden - I assume - due to a bug in the stubbing library and concurrent calls.

Reply:
Ah, thanks for the additional info. I'll experiment with that further.

Reply:
Ok, _I think_ I've got the test coverage working. I fixed my previous test to use the grep count of successes. I've added two new tests, the first tests that it will wait when the concurrency limit has been reached, the second tests that the concurrency loop will continue when there are curl failures.

I tried to add another test for concurrency timeouts but was struggling to get that to play nice with stub as it doesn't seem to like it when the underlying process is killed when unstub is called.

I also added a new var, specifically for enabling additional logs for test output... enabling debug breaks some tests... I figured it would be useful to have a new logging var specifically for increased verbosity while running tests.

Reply:
> Ok, _I think_ I've got the test coverage working. I fixed my previous test to use the grep count of successes. I've added two new tests, the first tests that it will wait when the concurrency limit has been reached, the second tests that the concurrency loop will continue when there are curl failures.

Ran the tests and they appear to be working :)

> I tried to add another test for concurrency timeouts but was struggling to get that to play nice with stub as it doesn't seem to like it when the underlying process is killed when unstub is called.

kind of makes sense as the stubbing process does a lot of recording in the background about what was called and what happened when it did so if those processes get killed it makes sense that they do not get recorded appropriately (like it happens with the concurrent executions) and unstubbing fails.

That said, I believe you could take care of everything by also stubbing the `kill` command so that the processes don't actually get killed.

Reply:
I had considered that but my worry was that stubbing `kill` would also stub the `kill -0` which has a different purpose. Do you know if there would be a way of doing one but not the other?

Reply:
Nope, if you stub a command you will need to mock all calls to it. Luckily the call to `kill -0` is quite simple and straight-forward, if the command run in the mock fails it will assume that the process has finished and remove it from the _waitlist_, otherwise it will assume it is alive and proceed accordingly. What's more, you only need to take care of all that in a single test, not all of them.

And remember to add the corresponding comment at the top for easier debugging in the future

Reply:
I think another bug with concurrency might be hampering this, I've added the following test:

```
@test "Concurrency gracefully handles command-group timeout" {
  export BUILDKITE_PLUGIN_TEST_COLLECTOR_FILES='**/*/junit-*.xml'
  export BUILDKITE_PLUGIN_TEST_COLLECTOR_UPLOAD_CONCURRENCY='2'
  export BUILDKITE_PLUGIN_TEST_COLLECTOR_TIMEOUT='3'

  stub kill \
    "-0 \*: echo 'check command-group' \
    "\*: echo 'killed command-group'"

  stub curl \
    "-X POST --silent --show-error --max-time 3 --form format=junit ${COMMON_CURL_OPTIONS} \* -H \* : sleep 60" \
    "-X POST --silent --show-error --max-time 3 --form format=junit ${COMMON_CURL_OPTIONS} \* -H \* : echo 'curl success'"

  run "$PWD/hooks/pre-exit"

  unstub kill
  unstub curl

  assert_success
  assert_output --partial "Uploading './tests/fixtures/junit-1.xml'..."
  assert_output --partial "Uploading './tests/fixtures/junit-2.xml'..."
  assert_equal "$(echo "$output" | grep -c "has been running for more than")" "2"
  assert_output --partial "Uploading './tests/fixtures/junit-3.xml'..."
  assert_equal "$(echo "$output" | grep -c "curl success")" "1"
}
```

However the test is failing with:

```
 ✗ Concurrency gracefully handles command-group timeout
   (from function `unstub' in file /usr/lib/bats/bats-mock/stub.bash, line 70,
    in test file tests/pre-exit-success.bats, line 204)
     `unstub kill' failed
   The stub for kill wasn't run
```

The `kill` is definitely being executed because it stops the sleep after around 3s. I've set `KILL_STUB_DEBUG` and it doesn't look like it ever attempts to stub kill, which makes sense given the test failure output. I'll commit my work on this so far in-case you want to have a poke.

Reply:
I didn't get a lot of time on this today, but I'm thinking perhaps we ditch this test for now... it would be useful to have but I can't get it to behave.

Reply:
So sorry for the delay on getting back to you and thanks a lot for your patience. I was sure this was completely possible and it was... after hours and hours of banging my head against the following:

* stubs with concurrent executions get weird (and need to be configured and treated weirdly)
* `kill` is a bash builtin so it can not be mocked (mea culpa, it is also a command)
* a stub that only does `sleep X` can not get killed not matter what (thanks bash)
* a stub that gets killed does not appear to cleanup after itself or record it was executed (adding to the already weird behaviour)

So, in a nutshell, your (now removed) test will be successful if you:
1- remove the `kill` stub that does nothing
2- change the `curl` stub to have a single execution with the following:

`stub curl "if [ \${10} != 'data=@\"./tests/fixtures/junit-3.xml\"' ]; then echo sleeping for \${10}; sleep 10 & wait \$!; else echo curl success \${10}; fi"`

How did I arrive at this magical incantation you ask?
* concurrent executions mean that instead of 3 stubs, only 2 are necessary (one for each batch)
* the first batch gets killed so it would appear as never executed, thus reducing the amount of stubs to just a single command
* the way to get different behaviour off a single command was to add an `if` clause to inspect the file being uploaded and do different things there
* for that situation when the process will get killed, `sleep` can not be used as it will not handle the signal itself, thus `sleep X & wait \$!` ([relevant information](http://mywiki.wooledge.org/SignalTrap#When_is_the_signal_handled.3F) thanks to [this stackoverflow answer](https://stackoverflow.com/a/27694965/1352026))

If you are still reading here, I profoundly apologize for downplaying the difficulty of getting this to work and solemnly swear that any further review of this PR from me will come with a complete and tested patch instead of vague suggestions of how things should be only to let you down the level of hell that is reserved for those of us that mostly enjoy programming in bash :imp:

Reply:
Thank you so much for the comprehensive explanation! I'll try that out tomorrow :)

> I profoundly apologize for downplaying the difficulty of getting this to work

No stress, it's helped me learn a lot about bats which I've never used before. Unfortunately my workload has been a bit higher the last couple of weeks and it's just been hard to find enough time to solve this particularly challenging test. I really appreciate the time you've spent on this to help and there's no expectation on my end for you to provide a complete solution but I definitely needed a hand on this one :sweat_smile:

Reply:
Alrighty, I think I have a working implementation! Unfortunately it looks like I need to check for "error, continuing" rather than the timeout message specifically... but afaict that might be unavoidable as messages outside of the scope of the command group seem to get lost.

Regardless, it still tests that the first group is failing and that we return only one success, which might be close enough for our purposes here.

Reply:
Actually, my tests suggest that nothing actually happens if the process is killed so you can not check for either output:

```
     Uploading '3' files matching '**/*/junit-*.xml'
     Uploading './tests/fixtures/junit-1.xml'...
     Uploading './tests/fixtures/junit-2.xml'...
     sleeping for data=@"./tests/fixtures/junit-1.xml"
     sleeping for data=@"./tests/fixtures/junit-2.xml"
     Upload '2678' has been running for more than '3' seconds, killing it
     Upload '2679' has been running for more than '3' seconds, killing it
     Uploading './tests/fixtures/junit-3.xml'...
     curl success data=@"./tests/fixtures/junit-3.xml"
```

But I believe it is testing that the process does not break explosively :)

Reply:
Huh, weird, because I was seeing:

```
Uploading '3' files matching '**/*/junit-*.xml'
Uploading './tests/fixtures/junit-1.xml'...
Uploading './tests/fixtures/junit-3.xml'...
sleeping for data=@"./tests/fixtures/junit-1.xml"
curl success data=@"./tests/fixtures/junit-3.xml"
/plugin/hooks/pre-exit: line 180: kill: (83) - No such process
Uploading './tests/fixtures/junit-2.xml'...
Error uploading, will continue
```

I wonder why we're seeing different output 🤔. Probably not a big deal, interesting though.

Reply:
When I run it, the first batch is `1` and `2`, the second batch is just `3` so the first two processes get killed. In your run, the first batch was `1` and `3`, only one of those gets killed so the second one generates the `No such process` line (which points to the fact that the command needs stderr redirection to `/dev/null` It may be related to how files are laid out in the filesystem for the order in which `find` reports them back.

Most interestingly: the last one fails because of the `curl` stub only having one line (in your scenario it would need two) and that causes the `Error uploading` line, otherwise you should be seeing `sleeping for ` on the `2` file as well.

Instead of trying to figure out exactly how to work around this non-determinism, I would suggest you add a `| sort` to the `FIND_CMD[@]` call in line 156 so that the order of files is always the same no matter what.

Reply:
Yep, that fixed that issue :)